flytekit/annotated/promise.py

```diff
@@ -362,23 +362,43 @@ def __str__(self):
 
 
 # To create a class that is a named tuple, we might have to create namedtuplemeta and manipulate the tuple
-def create_task_output(promises: Optional[Union[List[Promise], Promise]]) -> Optional[Union[Tuple[Promise], Promise]]:
+def create_task_output(
+    promises: Optional[Union[List[Promise], Promise]], entity_interface: Optional[Interface] = None
+) -> Optional[Union[Tuple[Promise], Promise]]:
     if promises is None:
         return None
 
     if isinstance(promises, Promise):
         return promises
 
     if len(promises) == 0:
-        return None
+        raise Exception("jfkdlsjafklsdajlksda")
+        # Making this an exception for now
+        # return None
 
     if len(promises) == 1:
-        return promises[0]
-
-    # More than one promises, let us wrap it into a tuple
+        if not entity_interface:
+            return promises[0]
+        # See transform_signature_to_interface for more information, we're using the existence of a name as a proxy
+        # for the user having specified a one-element typing.NamedTuple, which means we should _not_ extract it. We
+        # should still return a tuple but it should be one of ours.
+        if not entity_interface.custom_interface_name:
+            return promises[0]
+
+    # More than one promise, let us wrap it into a tuple
+    # Start with just the var names in the promises
     variables = [p.var for p in promises]
 
-    class Output(collections.namedtuple("TaskOutput", variables)):
+    # These should be OrderedDicts so it should be safe to do this.
+    if entity_interface:
+        variables = [k for k in entity_interface.outputs.keys()]
+
+    named_tuple_name = "DefaultNamedTupleOutput"
+    if entity_interface and entity_interface.custom_interface_name:
+        named_tuple_name = entity_interface.custom_interface_name
+
+    # Should this class be part of the Interface?
```

Comment:
@kumare3 what do you think about this?

Reply:
what do you mean part? like it is an inner class in interface? It has to be generated on the fly, because of the named tuple semantics

Reply:
Why does it have to be generated on the fly?

Reply:
https://github.com/lyft/flytekit/commit/d4c911a290d4ae7ece3e0985cba598eb68667c4a#diff-b4c0ad757d04f70c1eef4ff55d25fd37a9afc669857ac3c0a4ba72f0abacd602R50 is this broken in some way I'm not seeing?

Reply:
wdym, you are generating the class on the Fly right? For every instance `__init__` you will generate a new class, albeit the same name, but that does not matter. 
I was doing it at the create output time, but I guess this is fine too